src/platform/x86_64/vmx/vmx_exit.c

```diff
@@ -0,0 +1,219 @@
+	vmx_assert(comp);
+	return expended_process(regs, thd_curr, comp, cos_info, 0);
+}
+
+void
+vmx_exit_handler(struct vm_vcpu_shared_region *regs)
+{
+	u64_t reason, qualification, gla, gpa, inst_length, inst_info;
+	u8_t reason_nr;
+	int ret = 0;
+	struct cos_cpu_local_info *cos_info;
+	struct thread *thd_curr, *thd_exception_handler, *next;
+	struct vm_vcpu_shared_region *shared_region;
+	cos_info = cos_cpu_local_info();
+	thd_curr = thd_current(cos_info);
+
+	thd_curr->vcpu_ctx.vmcs.guest_msr_gs_base = msr_get(IA32_GS_BASE);
+	thd_curr->vcpu_ctx.vmcs.guest_msr_gskernel_base = msr_get(IA32_KERNEL_GSBASE);
+	thd_curr->vcpu_ctx.vmcs.guest_tsc_aux = msr_get(IA32_TSC_AUX);
+	thd_curr->vcpu_ctx.vmcs.guest_star = msr_get(IA32_STAR);
+	thd_curr->vcpu_ctx.vmcs.guest_lstar = msr_get(IA32_LSTAR);
+	thd_curr->vcpu_ctx.vmcs.guest_cstar = msr_get(IA32_CSTAR);
+	thd_curr->vcpu_ctx.vmcs.guest_fmask = msr_get(IA32_FMASK);
+	
+	msr_set(IA32_GS_BASE, thd_curr->vcpu_ctx.vmcs.host_msr_gs_base);
+	msr_set(IA32_KERNEL_GSBASE, thd_curr->vcpu_ctx.vmcs.host_msr_gskernel_base);
+	msr_set(IA32_TSC_AUX, thd_curr->vcpu_ctx.vmcs.host_tsc_aux);
+	msr_set(IA32_STAR, thd_curr->vcpu_ctx.vmcs.host_star);
+	msr_set(IA32_LSTAR, thd_curr->vcpu_ctx.vmcs.host_lstar);
+	msr_set(IA32_CSTAR, thd_curr->vcpu_ctx.vmcs.host_cstar);
+	msr_set(IA32_FMASK, thd_curr->vcpu_ctx.vmcs.host_fmask);
+
+	vmx_assert(cos_info);
+	vmx_assert(thd_curr && thd_curr->cpuid == get_cpuid());
+	vmx_assert(thd_curr->thd_type == THD_TYPE_VM);
+
+	shared_region = thd_curr->vm_vcpu_shared_region;
+	thd_exception_handler = thd_curr->exception_handler;
+	
+	reason = vmread(EXIT_REASON);
+	qualification = vmread(EXIT_QUALIFICATION);
+	gla = vmread(EXIT_GUEST_LINEAR_ADDRESS);
+	gpa = vmread(EXIT_GUEST_PHYSICAL_ADDRESS);
+	inst_length = vmread(EXIT_INSTRUCTION_LENGTH);
+	inst_info = vmread(EXIT_INSTRUCTION_INFORMATION);
+
+	if (unlikely(reason & (1 << 31))) {
+		/* TODO: Should not happen, but need to handle this case correctly later */
+		VMX_DEBUG("exit reason:%x(%lu)\n", reason, reason);
+		VMX_DEBUG("exit qualification:0x%x\n", qualification);
+		vmx_assert(0);
+	}
+
+	reason_nr = reason & 0xffff;
+	vmx_assert(reason_nr < MAX_VM_EXIT_REASONS);
+	VMX_DEBUG("VM thd: %u on core: %u get VM-exit (reason: ) on handler: %u\n", thd_curr->tid, cos_info->cpuid, reason_nr, thd_exception_handler->tid);
+
+	/* Share GPs with VMM */
+	memcpy(shared_region, regs, sizeof(struct vm_vcpu_shared_region));
+
+	shared_region->reason = reason_nr;
+	shared_region->ip = vmread(GUEST_RIP);
+	shared_region->sp = vmread(GUEST_RSP);
+	shared_region->efer = vmread(GUEST_IA32_EFER);
+	shared_region->cr0 = vmread(GUEST_CR0);
+	shared_region->cr4 = vmread(GUEST_CR4);
+	shared_region->interrupt_status = vmread(GUEST_INTERRUPT_STATUS);
+	shared_region->inst_length = inst_length;
+	shared_region->gpa = gpa;
+	shared_region->qualification = qualification;
+
+	/* FIXME: handle vcpu execution time correctly, this also seems should add into VMM thread */
```

Comment:
Tried that before, but I forget why that code didn't work. Probably because of there is already a scheduler there. For the initial vmx support, this version should be fine. Will add a TODO here.